### PR TITLE
Overlord v2 - Refactored based on feedback

### PR DESCRIPTION
**Pull request by dustinhansen**
bomb_spec.rb - Joe mentioned to stop using an instance variable for
@security_codes, but I need that variable to persist to the random_code
method. Using let(), it would lose scope every time. An instance
variable is the only way I’ve seen to do this and have found lots of
examples of people using it. If there’s a more elegant way I’d love for
someone to show me.

### CODE REVIEW

File: overlord/bomb.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
+class Bomb
+  attr_reader :error_count
+  attr_reader :status
+
+  def initialize(activation_code = '1234', deactivation_code = '0000')
+    @activation_code = activation_code
+    @deactivation_code = deactivation_code
+    @status = 'inactive'
```
Comment (incazteca) [inline overlord/bomb.rb]: Use symbols for setting statuses.

File: overlord/overlord.rb [hunk 2]
```diff
@@ -1,14 +1,39 @@
...
+require './bomb'
 
-enable :sessions
+class Overlord < Sinatra::Base
+  enable :sessions
 
-get '/' do
-  "Time to build an app around here. Start time: " + start_time
-end
+  get '/?:activation_code?/?:deactivation_code?' do
```
Comment (incazteca) [inline overlord/overlord.rb]: Make your routes meaningful. This should be a POST to setup with the activation and deactivation codes that a user will want to set up.

File: overlord/overlord.rb [hunk 3]
```diff
@@ -1,14 +1,39 @@
...
+  end
+
+  post '/security' do
+    submit_code(params[:code])
+    response = { bomb_status: session[:bomb].status }
+    response[:error_count] = session[:bomb].error_count
+    response.to_json
+  end
+
+  def create_bomb(params)
```
Comment (incazteca) [inline overlord/overlord.rb]: This entire method is not really necessary. The bomb can be initialized when a user enters the overlord page. You can initially create the bomb `session[:bomb] = Bomb.new` and then have a setup route to set the activation codes with a `set_activation_code` method. Or you could check if the parameters given are present and then pass them in when creating the bomb.

File: overlord/views/index.erb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+<html>
+	<head>
+		<title>Overlord</title>
+		<link href="/assets/lib/bootstrap/css/bootstrap.min.css" rel="stylesheet">
+		<link href="/assets/app/css/main.css" rel="stylesheet">
+	</head>
+	<body>
+		<h1>Overlord Example</h1>
+		<form name="security" action="/" method="post">
+			<input type="number" name="code" max="9999">
```
Comment (incazteca) [inline overlord/views/index.erb]: Provide a box for the activation code and deactivation code. The current page is very confusing. Also include a label for the input.

File: overlord/spec/bomb_spec.rb [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+  def deactivate_bomb
+    subject.deactivate(@security_codes[:deactivate])
+  end
+
+  def explode_bomb
+    activate_bomb
+    3.times { subject.deactivate(random_code) }
+  end
+
+  before(:all) do
```
Comment (incazteca) [inline overlord/spec/bomb_spec.rb]: Use let blocks for the code

```
let(good_activation_code) {'1234'}
let(bad_code) {'abcd'}
etc....
```

File: overlord/features/security_codes.feature [hunk 6]
```diff
@@ -0,0 +1,39 @@
+Feature: Security Codes
```
Comment (incazteca) [inline overlord/features/security_codes.feature]: Your feature files should be grouped by a type of activity. Such as activation or deactivation of a bomb. An activation feature file will then include steps that are related to activating a bomb.

File: overlord/features/status_indicator.feature [hunk 7]
```diff
@@ -0,0 +1,19 @@
+Feature: Status Indicator
```
Comment (incazteca) [inline overlord/features/status_indicator.feature]: You should just use the status indicator in the Then steps in the other feature files. There's no real need for these tests.

**Comment by incazteca** [comment]
The bomb spec needs some work in using the let blocks as well as removing the loop at the bottom. Also the features files should be better organized. Also the routing for this needs to be worked on as well so that they are meaningful. Please resubmit.

File: overlord/public/assets/lib/bootstrap/js/bootstrap.js [hunk 8]
```diff
@@ -0,0 +1,2317 @@
+/*!
```
Comment (djkotowski) [inline overlord/public/assets/lib/bootstrap/js/bootstrap.js]: This should be minified.

File: overlord/features/security_codes.feature [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+  In order to securely arm or disarm the bomb
+  As a bomber
+  I want to be able to enter security codes that change the status of the bomb
+
+  @javascript
+  Scenario: Deactivate an inactive bomb
+    Given I am viewing an inactive bomb
+    When I enter any code that is not the activation code 1 time
```
Comment (djkotowski) [inline overlord/features/security_codes.feature]: This is kind of a long way of just saying "When I enter an invalid activation code _n_ times".  (I'm being very nitpicky, here.  Not a big deal.)

File: overlord/features/security_codes.feature [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+  @javascript
+  Scenario: Unsuccessfully deactivate an active bomb
+    Given I am viewing an active bomb
+    When I enter any code that is not the activation code 3 times
+    Then the bomb is exploded
+
+  @javascript
+  Scenario: Interacting with an exploded bomb
+    Given I am viewing an exploded bomb
+    Then the security code form should be gone
```
Comment (djkotowski) [inline overlord/features/security_codes.feature]: You generally will not want a `Given`, `Then` scenario without a `When`.  In this case, the action `When` is viewing the exploded bomb.

File: overlord/features/step_definitions/bomb_steps.rb [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    submit_security_form
+  end
+
+  def default_activation_code
+    '1234'
+  end
+
+  def default_deactivation_code
+    '0000'
+  end
```
Comment (djkotowski) [inline overlord/features/step_definitions/bomb_steps.rb]: The defaults should be constants.

File: overlord/features/step_definitions/bomb_steps.rb [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+      4.times { code += rand(9).to_s }
+      return code unless blacklist.include?(code)
+    end
+  end
+
+  def submit_security_form
+    find('form[name="security"] input[type="submit"]').click
+  end
+end
+World(SessionStepsHarness)
```
Comment (djkotowski) [inline overlord/features/step_definitions/bomb_steps.rb]: Why define these in a module only to bring the module into the world?  You can just extract these methods out.

File: overlord/spec/bomb_spec.rb [hunk 13]
```diff
@@ -0,0 +1,122 @@
+require 'pry'
+require './bomb.rb'
+
+describe Bomb do
+  def random_code(num = 4)
+    loop do
+      code = ''
+      num.times { code += rand(9).to_s }
+      return code unless @security_codes.include? code
+    end
```
Comment (djkotowski) [inline overlord/spec/bomb_spec.rb]: This is flappy, as you'll be using a different code each time.  Think of the edge cases and explicitly define the codes you want entered.

File: overlord/spec/bomb_spec.rb [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+      it 'should remain in "exploded" state when we enter the correct deactivation code' do
+        subject.deactivate(random_code)
+        expect(subject.status).to eq('exploded')
+      end
+    end
+  end
+
+  context 'using default security codes' do
+    it_behaves_like 'mutable bomb' do
+    end
```
Comment (djkotowski) [inline overlord/spec/bomb_spec.rb]: Remove the empty code block.

**Comment by djkotowski** [comment]
Good work!  Take a look at the comments and refactor.  Resubmit when ready.

**Comment by dustinhansen** [comment]
I am still having a really tough time understanding what the ask is for creating Bombs. We're not using a database so I have no way of storing information securely. My only option is to store the activation/deactivation codes in a cookie. Is that acceptable? I'm dying for a database right now.

**Pull request closed without merging**